File: protocol-autoresearch/experiments/tier2_005_sdk_designer_proof_envelope_codec/sdk-python/bolyra/envelope.py
"""
BolyraEnvelope — self-describing proof envelope with named public signals.

Python mirror of sdk/src/envelope.ts.  Serialises/deserialises via standard
json with string-encoded bigints (no Python-specific types in the wire format).
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from .signals import SIGNAL_MAPS, VALID_CIRCUITS, VALID_PROVING_SYSTEMS

# ---------------------------------------------------------------------------
# Version
# ---------------------------------------------------------------------------

ENVELOPE_VERSION = "1.0.0"
_SUPPORTED_MAJOR = 1


# ---------------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------------


class EnvelopeError(Exception):
    """Raised on encode/decode validation failures."""

    def __init__(self, code: str, message: str) -> None:
        self.code = code
        super().__init__(f"[{code}] {message}")


# ---------------------------------------------------------------------------
# Dataclass
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class BolyraEnvelope:
    """Self-describing proof envelope."""

    version: str
    circuit: str
    proving_system: str
    signals: Dict[str, str]
    proof: Dict[str, Any]

    # -- JSON serialisation --------------------------------------------------

    def to_json(self) -> str:
        """Serialise to canonical JSON (camelCase keys on the wire)."""
        return json.dumps(
            {
                "version": self.version,
                "circuit": self.circuit,
                "provingSystem": self.proving_system,
                "signals": self.signals,
                "proof": self.proof,
            },
            separators=(",", ":"),
        )

    def to_dict(self) -> Dict[str, Any]:
        """Return a JSON-compatible dict (camelCase keys)."""
        return {
            "version": self.version,
            "circuit": self.circuit,
            "provingSystem": self.proving_system,
            "signals": dict(self.signals),
            "proof": dict(self.proof),
        }

    @staticmethod
    def from_json(raw: str) -> "BolyraEnvelope":
        """Deserialise from JSON, validating version and circuit."""
        try:
            data = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise EnvelopeError("INVALID_JSON", f"Cannot parse JSON: {exc}") from exc
        return BolyraEnvelope.from_dict(data)

    @staticmethod
    def from_dict(data: Dict[str, Any]) -> "BolyraEnvelope":
        """Construct from a dict (camelCase keys), validating fields."""
        version = data.get("version", "")
        _validate_version(version)

        circuit = data.get("circuit", "")
        if circuit not in VALID_CIRCUITS:
            raise EnvelopeError(
                "UNKNOWN_CIRCUIT",
                f'Unknown circuit "{circuit}". '
                f"Expected one of: {', '.join(sorted(VALID_CIRCUITS))}",
            )

        proving_system = data.get("provingSystem", "")
        if proving_system not in VALID_PROVING_SYSTEMS:
            raise EnvelopeError(
                "UNKNOWN_PROVING_SYSTEM",
                f'Unknown proving system "{proving_system}". '
                f"Expected one of: groth16, plonk",
            )

        signals = data.get("signals", {})
        expected_names = set(SIGNAL_MAPS[circuit])
        actual_names = set(signals.keys())
        if actual_names != expected_names:
            missing = expected_names - actual_names
            extra = actual_names - expected_names
            parts = []
            if missing:
                parts.append(f"missing={missing}")
            if extra:
                parts.append(f"extra={extra}")
            raise EnvelopeError(
                "SIGNAL_MISMATCH",
                f"Signal keys do not match {circuit}: {'; '.join(parts)}",
            )

        return BolyraEnvelope(
            version=version,
            circuit=circuit,
            proving_system=proving_system,
            signals={k: str(v) for k, v in signals.items()},
            proof=data.get("proof", {}),
        )


# ---------------------------------------------------------------------------
# encode / decode / from_raw
# ---------------------------------------------------------------------------


def encode(
    circuit: str,
    proving_system: str,
    raw_proof: Dict[str, Any],
    raw_signals: List[str],
) -> BolyraEnvelope:
    """Encode raw snarkjs output into a BolyraEnvelope."""
    if circuit not in VALID_CIRCUITS:
        raise EnvelopeError(
            "UNKNOWN_CIRCUIT",
            f'Unknown circuit "{circuit}". '
            f"Expected one of: {', '.join(sorted(VALID_CIRCUITS))}",
        )
    if proving_system not in VALID_PROVING_SYSTEMS:
        raise EnvelopeError(
            "UNKNOWN_PROVING_SYSTEM",
            f'Unknown proving system "{proving_system}". Expected one of: groth16, plonk',
        )

    names = SIGNAL_MAPS[circuit]
    if len(raw_signals) != len(names):
        raise EnvelopeError(
            "SIGNAL_COUNT_MISMATCH",
            f"Expected {len(names)} signals for {circuit}, got {len(raw_signals)}",
        )

    signals = {names[i]: str(raw_signals[i]) for i in range(len(names))}

    return BolyraEnvelope(
        version=ENVELOPE_VERSION,
        circuit=circuit,
        proving_system=proving_system,
        signals=signals,
        proof=raw_proof,
    )


def decode(envelope: BolyraEnvelope) -> Dict[str, Any]:
    """Decode a BolyraEnvelope back to positional { proof, publicSignals }."""
    _validate_version(envelope.version)

    if envelope.circuit not in VALID_CIRCUITS:
        raise EnvelopeError(
            "UNKNOWN_CIRCUIT",
            f'Unknown circuit "{envelope.circuit}".',
        )

    names = SIGNAL_MAPS[envelope.circuit]
    public_signals: List[str] = []
    for name in names:
        val = envelope.signals.get(name)
        if val is None:
            raise EnvelopeError(
                "MISSING_SIGNAL",
                f'Signal "{name}" missing from envelope for circuit {envelope.circuit}',
            )
        public_signals.append(str(val))

    return {"proof": envelope.proof, "publicSignals": public_signals}


def from_raw(
    circuit: str,
    proving_system: str,
    raw_proof: Dict[str, Any],
    raw_signals: List[str],
) -> BolyraEnvelope:
    """Convenience alias for encode() — migration entry point."""
    return encode(circuit, proving_system, raw_proof, raw_signals)


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _validate_version(version: str) -> None:
    """Validate that version string has a supported major version."""
    try:
        major = int(version.split(".")[0])
    except (ValueError, IndexError) as exc:
        raise EnvelopeError(
            "INVALID_VERSION", f"Cannot parse version: {version}"
        ) from exc

    if major != _SUPPORTED_MAJOR:
        raise EnvelopeError(
            "UNSUPPORTED_VERSION",
            f"Envelope version {version} (major {major}) is not supported. "
            f"Expected major {_SUPPORTED_MAJOR}.",
        )

File: protocol-autoresearch/experiments/tier2_005_sdk_designer_proof_envelope_codec/sdk/src/envelope.ts
/**
 * BolyraEnvelope — self-describing proof envelope with named public signals.
 *
 * Replaces bare `{ proof, publicSignals: bigint[] }` with a versioned,
 * circuit-aware envelope that maps positional signals to human-readable names.
 *
 * @module
 */

import {
  SIGNAL_MAPS,
  VALID_CIRCUITS,
  VALID_PROVING_SYSTEMS,
  type BolyraCircuit,
  type BolyraProvingSystem,
} from './signals.js';

// ---------------------------------------------------------------------------
// Version
// ---------------------------------------------------------------------------

/** Current envelope schema version. */
export const ENVELOPE_VERSION = '1.0.0';

/** Supported major version for decode validation. */
const SUPPORTED_MAJOR = 1;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** snarkjs proof object (passthrough). */
export interface SnarkjsProof {
  readonly pi_a: readonly string[];
  readonly pi_b: readonly (readonly string[])[];
  readonly pi_c: readonly string[];
  readonly protocol: string;
  readonly curve: string;
}

/** Named public signals keyed by field name. Values are string-encoded bigints. */
export type NamedSignals = Record<string, string>;

/** Self-describing proof envelope. */
export interface BolyraEnvelope {
  readonly version: string;
  readonly circuit: BolyraCircuit;
  readonly provingSystem: BolyraProvingSystem;
  readonly signals: NamedSignals;
  readonly proof: SnarkjsProof;
}

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

export class EnvelopeError extends Error {
  constructor(public readonly code: string, message: string) {
    super(`[${code}] ${message}`);
    this.name = 'EnvelopeError';
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function parseMajor(version: string): number {
  const n = parseInt(version.split('.')[0], 10);
  if (Number.isNaN(n)) {
    throw new EnvelopeError('INVALID_VERSION', `Cannot parse version: ${version}`);
  }
  return n;
}

function getSignalNames(circuit: string): readonly string[] {
  if (!VALID_CIRCUITS.has(circuit)) {
    throw new EnvelopeError(
      'UNKNOWN_CIRCUIT',
      `Unknown circuit "${circuit}". Expected one of: ${[...VALID_CIRCUITS].join(', ')}`,
    );
  }
  return SIGNAL_MAPS[circuit as BolyraCircuit];
}

// ---------------------------------------------------------------------------
// encode
// ---------------------------------------------------------------------------

/**
 * Encode raw snarkjs output into a BolyraEnvelope.
 *
 * Zips the positional `rawSignals` array with the signal map for `circuit`
 * to produce a named `signals` record.
 *
 * @param circuit       - Circuit that produced the proof
 * @param provingSystem - Proving system used (groth16 | plonk)
 * @param rawProof      - snarkjs proof object
 * @param rawSignals    - snarkjs publicSignals array (positional bigint strings)
 * @returns A fully populated BolyraEnvelope
 */
export function encode(
  circuit: BolyraCircuit,
  provingSystem: BolyraProvingSystem,
  rawProof: SnarkjsProof,
  rawSignals: readonly string[],
): BolyraEnvelope {
  if (!VALID_PROVING_SYSTEMS.has(provingSystem)) {
    throw new EnvelopeError(
      'UNKNOWN_PROVING_SYSTEM',
      `Unknown proving system "${provingSystem}". Expected one of: groth16, plonk`,
    );
  }

  const names = getSignalNames(circuit);

  if (rawSignals.length !== names.length) {
    throw new EnvelopeError(
      'SIGNAL_COUNT_MISMATCH',
      `Expected ${names.length} signals for ${circuit}, got ${rawSignals.length}`,
    );
  }

  const signals: Record<string, string> = {};
  for (let i = 0; i < names.length; i++) {
    signals[names[i]] = String(rawSignals[i]);
  }

  return {
    version: ENVELOPE_VERSION,
    circuit,
    provingSystem,
    signals,
    proof: rawProof,
  };
}

// ---------------------------------------------------------------------------
// decode
// ---------------------------------------------------------------------------

/**
 * Decode a BolyraEnvelope back to positional `{ proof, publicSignals }`
 * suitable for `snarkjs.groth16.verify()` / `snarkjs.plonk.verify()`.
 *
 * @param envelope - A BolyraEnvelope (e.g. received over the wire)
 * @returns Object with `proof` and `publicSignals` (positional string array)
 * @throws EnvelopeError on version mismatch or unknown circuit
 */
export function decode(envelope: BolyraEnvelope): {
  proof: SnarkjsProof;
  publicSignals: string[];
} {
  // Version gate
  const major = parseMajor(envelope.version);
  if (major !== SUPPORTED_MAJOR) {
    throw new EnvelopeError(
      'UNSUPPORTED_VERSION',
      `Envelope version ${envelope.version} (major ${major}) is not supported. Expected major ${SUPPORTED_MAJOR}.`,
    );
  }

  const names = getSignalNames(envelope.circuit);
  const publicSignals: string[] = new Array(names.length);

  for (let i = 0; i < names.length; i++) {
    const val = envelope.signals[names[i]];
    if (val === undefined) {
      throw new EnvelopeError(
        'MISSING_SIGNAL',
        `Signal "${names[i]}" missing from envelope for circuit ${envelope.circuit}`,
      );
    }
    publicSignals[i] = String(val);
  }

  return { proof: envelope.proof, publicSignals };
}

// ---------------------------------------------------------------------------
// fromRaw (migration convenience)
// ---------------------------------------------------------------------------

/**
 * Convenience alias for `encode()` — entry point for migrating existing
 * callers from bare `{ proof, publicSignals }` to envelopes.
 *
 * @example
 * ```ts
 * // Before (positional — error-prone):
 * const root = publicSignals[0]; // wait, is it [0] or [2]?
 *
 * // After (named — self-documenting):
 * const env = fromRaw('HumanUniqueness', 'groth16', proof, publicSignals);
 * const root = env.signals.humanMerkleRoot; // ✓
 * ```
 */
export function fromRaw(
  circuit: BolyraCircuit,
  provingSystem: BolyraProvingSystem,
  rawProof: SnarkjsProof,
  rawSignals: readonly string[],
): BolyraEnvelope {
  return encode(circuit, provingSystem, rawProof, rawSignals);
}
